sno/dataset2.py

```diff
@@ -122,6 +125,141 @@ def hexhash(self):
+            **self.extra_type_info,
+        }
+
+    def __eq__(self, other):
+        if not isinstance(other, ColumnSchema):
+            return False
+        return (
+            self.id == other.id
+            and self.name == other.name
+            and self.data_type == other.data_type
+            and self.pk_index == other.pk_index
+            and self.extra_type_info == other.extra_type_info
+        )
+
+
+class Schema:
+    """A schema is just an ordered list of ColumnSchemas."""
+
+    def __init__(self, columns):
+        """Create a new schema from a list of ColumnSchemas."""
+        self.columns = tuple(columns)
+        # Creating a legend validates the primaryKeyIndex field.
+        self.legend = self._to_legend()
+
+    def __getitem__(self, i):
+        """Return the _i_th ColumnSchema."""
+        return self.columns[i]
+
+    @classmethod
+    def loads(cls, data):
+        """Load a schema from a bytestring"""
+        json_array = _unjson(data)
+        columns = [ColumnSchema.from_json_dict(c) for c in json_array]
+        return cls(columns)
+
+    def dumps(self):
+        """Writes this schema to a bytestring."""
+        json_array = [c.to_json_dict() for c in self.columns]
+        return _json(json_array)
+
+    def feature_dict_from_raw(self, raw_dict):
+        """Values keyed by column ID -> values keyed by column name."""
+        feature_dict = {}
+        for column in self.columns:
+            feature_dict[column.name] = raw_dict.get(column.id, None)
+        return feature_dict
+
+    def feature_tuple_from_raw(self, raw_dict):
+        """Values keyed by column ID -> value tuple in schema order."""
+        feature_tuple = []
+        for column in self.columns:
+            feature_tuple.append(raw_dict.get(column.id, None))
+        return tuple(feature_tuple)
+
+    def feature_dict_to_raw(self, feature_dict):
+        """Values keyed by column name -> values keyed by column ID."""
+        raw_dict = {}
+        for column in self.columns:
+            raw_dict[column.id] = feature_dict[column.name]
+        return raw_dict
+
+    def feature_tuple_to_raw(self, feature_tuple):
+        """Value tuple in schema order -> values keyed by column ID."""
+        assert len(feature_tuple) == len(self.columns)
+        raw_dict = {}
+        for column, value in zip(self.columns, feature_tuple):
+            raw_dict[column.id] = value
+        return raw_dict
+
+    def _to_legend(self):
+        pk_column_ids = []
+        non_pk_column_ids = []
+        for i, column in enumerate(sorted(self.columns, key=_pk_index_ordering)):
+            if column.pk_index is not None:
+                if i != column.pk_index:
+                    raise ValueError(
+                        f"Expected contiguous primaryKeyIndex {i} but only found {column.pk_index}"
+                    )
+                pk_column_ids.append(column.id)
+            else:
+                non_pk_column_ids.append(column.id)
```

Comment:
Doesn't this mean non-PK columns end up in a random order?

My understanding is we want the legend/blob to be sorted by column index, but the PK column values aren't actually present in the feature blob (because they're in the tree path).

Reply:
hmm, reading back that doesn't make a huge lot of sense. 😉

Reply:
Something needs to put the columns in the right order for working copies/exports/whatever — seems like it should be the Dataset class responsibility?

Reply:
Theory: because python sort is stable, this sort leaves non-PK columns stay in their row order. Practice: There is a test that correctly roundtrips a feature in spite of the schema changing column order and in spite of the feature dictionaries provided having the fields initialised in random order.

Reply:
Dataset2.read_feature_tuple returns the row in row order
Dataset2.read_feature_dict returns a dict that has been initialised in row order, so that iterating over stays in row order

Schema.feature_tuple_from_raw and Schema.feature_dict_from_raw are where the data is put into row order - they iterate through schema.columns and add data for each column in order. (They might need to be optimised one day, every feature read creates a temporary dict)

test_schema_change_roundtrip shows that reading a feature gives the feature in row-order, even if the row order is changed due to a new schema.